### PR TITLE
Auth with API Key

### LINKED ISSUE #36570: Write a middleware to allow the use of API keys for authentication

### PR DESCRIPTION
**Pull request by johnswanson**
In terms of the actual application code, things are pretty straightforward: if an API key is passed, find the associated user, check the hash against the passed API key, and attach the user if everything went well.

I renamed the existing functions related to API keys, e.g. `mw.auth/wrap-api-key`. We should be careful about overloading the term "API Key", both internally and externally.

(The "API Key" for those existing functions is a static key defined by an env var that allows you to require a particular header *in addition to regular authentication* on some endpoints.)

It's probably too painful to change the actual setting name, but we can change what we call it in the code.

Fixes #36570

### CODE REVIEW

File: test/metabase/api/api_key_test.clj [hunk 1]
```diff
@@ -102,3 +103,14 @@
                             :user_id      (mt/user->id :crowberto)
                             :created_by   (mt/user->id :crowberto)}]
         (is (= 2 (mt/user-http-request :crowberto :get 200 "api-key/count")))))))
+
+(deftest api-keys-work-e2e
```
Comment (noahmoss) [inline test/metabase/api/api_key_test.clj:107]: Could you add a test that all of the session-level dynamic vars (like `api/is-superuser?`, etc) are bound correctly for API users?

File: src/metabase/server/middleware/session.clj [hunk 2]
```diff
@@ -305,17 +324,41 @@
...
 
+(def ^:private api-key-that-should-never-match (str (random-uuid)))
+(def ^:private hash-that-should-never-match (u.password/hash-bcrypt "password"))
+
+(defn- do-useless-hash []
+  (u.password/verify-password api-key-that-should-never-match "" hash-that-should-never-match))
+
+(defn- matching-api-key? [{:keys [api-key] :as _user-data} passed-api-key]
+  ;; if we get an API key, check the hash against the passed value. If not, don't reveal info via a timing attack - do
+  ;; a useless hash, *then* return `false`.
```
Comment (noahmoss) [inline src/metabase/server/middleware/session.clj:335]: awesome

File: test/metabase/server/middleware/session_test.clj [hunk 3]
```diff
@@ -243,6 +243,25 @@
               :uri                 "/anyurl"}
              (select-keys (wrapped-handler request) [:anti-csrf-token :cookies :metabase-session-id :uri]))))))
 
+(deftest current-user-info-for-api-key-test
+  (try
+    (t2/insert! :model/ApiKey {:id 999
+                               :name "An API Key"
+                               :user_id (mt/user->id :lucky)
+                               :created_by (mt/user->id :lucky)
+                               :unhashed_key "mb_foobar"})
```
Comment (noahmoss) [inline test/metabase/server/middleware/session_test.clj:252]: Can you use `t2.with-temp` here instead?

**Review by noahmoss — CHANGES_REQUESTED** [review]

File: src/metabase/server/middleware/session.clj [hunk 4]
```diff
@@ -290,6 +292,23 @@
...
+
+;; See above: because this query runs on every single API request (with an API Key) it's worth it to optimize it a bit
+;; and only compile it to SQL once rather than every time
+(def ^:private user-data-for-api-key-prefix-query
+  (first
+   (t2.pipeline/compile*
+    {:select    [[:api_key.user_id :metabase-user-id]
+                 [:api_key.key :api-key]
+                 [:user.is_superuser :is-superuser?]
+                 [:user.locale :user-locale]]
```
[the remark this answers is not in the document]
Comment (johnswanson) [reply src/metabase/server/middleware/session.clj:304]: Good point - this actually raises a question: can API key users become group managers, and if so, how? My thinking was that API key users will be excluded from the users you get back from normal user API endpoints, so you wouldn't be able to go through the normal flow for making them a group manager. But, that's maybe orthogonal to the question of "if a API key user _is_ a group manager, do we respect that?" - which I think the answer is yes, we should. I'll change this.

(And then maybe at some point down the road we'll add the UI/API necessary to actually make an API key into a group manager.)

**Review by noahmoss — APPROVED** [review]

File: src/metabase/server/middleware/session.clj [hunk 5]
```diff
@@ -305,17 +333,43 @@
...
+  ;; if we get an API key, check the hash against the passed value. If not, don't reveal info via a timing attack - do
+  ;; a useless hash, *then* return `false`.
+  (if api-key
+    (u.password/verify-password passed-api-key "" api-key)
+    (do-useless-hash)))
+
+(defn- current-user-info-for-api-key
+  "Return User ID and superuser status for an API Key with `api-key-id"
+  [api-key]
+  (when api-key
```
[the remark this answers is not in the document]
Comment (johnswanson) [reply src/metabase/server/middleware/session.clj]: Good call, done!

File: src/metabase/server/middleware/session.clj [hunk 6]
```diff
@@ -305,17 +333,43 @@
...
+      (when (matching-api-key? user-data api-key)
+        (dissoc user-data :api-key)))))
+
 (defn- merge-current-user-info
-  [{:keys [metabase-session-id anti-csrf-token], {:strs [x-metabase-locale]} :headers, :as request}]
+  [{:keys [metabase-session-id anti-csrf-token], {:strs [x-metabase-locale x-api-key]} :headers, :as request}]
   (merge
    request
    (current-user-info-for-session metabase-session-id anti-csrf-token)
+   (current-user-info-for-api-key x-api-key)
```
[the remark this answers is not in the document]
Comment (johnswanson) [reply src/metabase/server/middleware/session.clj]: Good idea, done - I used an `or` so we'll merge with the first non-nil user-info, if any (with the session taking precedence over the API key).

**Pull request merged**